archive/imo/imo1964_q1.lean

```diff
@@ -0,0 +1,93 @@
+/-
+Copyright (c) 2020 Kenji Nakagawa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kenji Nakagawa
+-/
+
+import field_theory.finite.basic
+
+lemma modfact (n : ℕ) : (2^n : zmod 7) = 2 ^ (n : zmod 3).val :=
+begin
+  have h1 : ∀ (k : ℕ), k < 6 → (2^k : zmod 7) = 2 ^ (k : zmod 3).val := by dec_trivial,
+  haveI : fact (nat.prime 7) := by { delta fact, norm_num },
+  have h2 : (2 : zmod 7) ≠ 0 := dec_trivial,
+  rw ← nat.mod_add_div n 6,
+  simp only [h2, pow_add, pow_mul, bit0_zero, one_pow, add_zero, mul_one, zmod.cast_self,
+    zmod.pow_bit0, nat.cast_bit0, zero_mul, ne.def, nat.cast_add, not_false_iff, nat.cast_mul],
+  apply h1,
+  apply nat.mod_lt,
+  norm_num,
+end
+
+/- TODO(?) : Refactor to only use `%` -/
+lemma modfact' (n : ℕ) : (2^n : ℕ) % 7 = 2^(n % 3) % 7 :=
+begin
+  have h1 : ∀(k : ℕ), k < 6 → ((2^k : ℕ) % 7) = 2^(k % 3) % 7 := by dec_trivial,
+  haveI : fact (nat.prime 7) := by { delta fact, norm_num },
+  have h2 : (2 % 7) ≠ 0 := dec_trivial,
+  rw ← nat.mod_add_div n 6,
+  have : 6 * (n / 6) % 3 = 0, sorry,
```

Comment:
Do you need help with these `sorry`s? Or are you working on them?

Reply:
I'm working on them, since the past few days have been rather busy.